bin/node/runtime/src/lib.rs

```diff
@@ -624,9 +628,85 @@ parameter_types! {
 	pub const StakingUnsignedPriority: TransactionPriority = TransactionPriority::max_value() / 2;
 }
 
+#[derive(Default)]
+pub struct ExtrasParams {
+	era: Option<Era>,
+	index: Option<Index>,
+	tip: u128,
+	prior_block_hash: Option<Hash>,
+}
+
+impl SystemExtraParams<Runtime> for ExtrasParams {
+	fn set_nonce(&mut self, index: Index) {
+		self.index = Some(index);
+	}
+
+	fn set_era(&mut self, era: Era) {
+		self.era = Some(era);
+	}
+
+	fn set_prior_block_hash(&mut self, hash: Hash) {
+		self.prior_block_hash = Some(hash);
+	}
+}
+
 
-impl<LocalCall> frame_system::offchain::CreateSignedTransaction<LocalCall> for Runtime where
-	Call: From<LocalCall>,
+impl SignedExtensionProvider for Runtime {
+	type Extra = SignedExtra;
+	type Params = ExtrasParams;
+
+	fn extension_params() -> ExtrasParams {
+		ExtrasParams {
+			prior_block_hash: None,
+			index: None,
+			era: None,
+			tip: 0,
+		}
+	}
+
+	fn construct_extras(params: ExtrasParams) -> Result<SignedExtensionData<Self::Extra>, &'static str> {
+		let ExtrasParams {
+			tip,
+			era,
+			index,
+			prior_block_hash
+		} = params;
+		let (era, index) = (
+			era.ok_or("era is required")?,
```

Comment:
IMHO that's not a good API design, would be better to ensure with type that these parameters are set or use some sensible defaults (i.e. immortal era, and from-state index)

Reply:
@tomusdrw can you elaborate? Do you mean they shouldn't error if they are missing but have a reasonable default? Or do you think the API of having an associated input-params is bad? If so, what would be the proposed change?

Reply:
Yeah, sorry for the random comment, with no clarity. I was skimming through the PR and noticed this.

The main reason I raised the comment is because I think having a generic error that indicates malformed input parameters in a function called `construct_extras` is a bit hard to reason about. If you just look at the types in the trait it seems okay to do:
```rust
Runtime::construct_extras(Runtime::extension_params())
``` 
and I'd assume that the "construction" of extras failed, not that my parameters are malformed, hence most likely I'd just propagate this error up instead of handling it, so clearly my own logical mistake is not caught by the type system, but rather would only pop up in the runtime.

I think there are two different paths that we can take:
1. Provide defaults inside `construct_extras` or
2. Separate `ExtrasParams` into a `Builder` and constructed struct.

In both cases `construct_extras` becomes infallible.

I think I'd prefer option (2) even though it will require more code, because (1) is:
- not explicit, 
- might become more difficult to maintain in the future,
- or the usage will end up being confusing (by adding more defaults)

With option (2) I imagine that `type Params` would be traitified with something like:
```rust
trait Params {
  trait Builder;
}
trait SignedExtensionProvider {
fn extension_params() -> Self::Params::Builder { .. }
fn construct_extras(params: Self::Params)...
} 
```
Also since the struct now becomes a builder the `fn set_*` methods might now return `&mut Self` which would allow for nice chaining.

Builder would be terminated with:
```rust
fn build(self) -> Result<Params, ExplicitNontStringBuilderError>;
// or maybe
fn build_with_defaults(self) -> Params;
 ```

That makes the error explicit to the builder and it's params - now it's a bit unclear if `construct_extras` fail because of malformed input or some other internal state. 
This forces the caller to handle the errors before the `construct_extras` can be called.

Hope it makes more sense :)